### PR TITLE
New Metadata Management

### PR DESCRIPTION
**Pull request by matthewpeterkort**
See https://github.com/calypr/forge/issues/2

A-lot of cleanup in the fhir_import_export job file for error handling / logging

TODO: confirm that ID minting strategies used in thei fhir_import_export job conform with existing mint id strategies used in existing ETL scripts.

**Edit (QW)**: Also see https://github.com/ACED-IDP/submission/pull/36 to patch pip install of `submission`

Testing:  run publish job  on  git-drs projects that have:
- metadata but not files
- files but no metadata
- metadata and files
- metadata  with different indexd information than the files

Observe no errors and that instances where there exists conflicting indexd file metadata with what exists in the uploaded document references, the indexd record metadata is used over what was uploaded as the META/DocumentReference.

Verify that inserted document references contain a reference to a research study.

### CODE REVIEW

File: etl-job/Dockerfile [hunk 1]
```diff
@@ -1,50 +1,44 @@
-# set base image (host OS)
+FROM golang:1.24.0-bookworm AS go-builder
+
+RUN go install github.com/calypr/git-drs@526654610a78fdd587aa66afc1e3dc22b0c155c4
```
Comment (bwalsh) [inline etl-job/Dockerfile]: should this be a tag or branch?

File: etl-job/fhir_import_export.py [hunk 2]
```diff
@@ -221,157 +207,449 @@ def _load_all(hostname,
...
+        'ghCommitHash': "input data must contain a `ghCommitHash`",
+        'ghRepoUrl': "input data must contain a `ghRepoUrl`",
+        'bucketName': "input data must contain a `bucketName`",
+        'profile': "input data must contain a `profile`",
+        'APIEndpoint': "input data must contain a `APIEndpoint`",
+    }
+
+    for field, error_msg in required_fields.items():
+        if field not in input_data or not input_data[field]:
+            raise ValueError(error_msg)
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: if  you wanna keep the same error message, you can shorten it

```suggestion
    required_fields = ['ghUserName', 'ghToken', 'ghCommitHash', 'ghRepoUrl', 'bucketName', 'profile', 'APIEndpoint', ]

    for field, error_msg in required_fields.items():
        if field not in input_data or not input_data[field]:
            raise ValueError(f"input data must contain a `{field}`")
```

File: etl-job/fhir_import_export.py [hunk 3]
```diff
@@ -221,157 +207,449 @@ def _load_all(hostname,
...
+
+    files = input_data['file']
+    if len(files) > 0:
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: same vibe here

File: etl-job/fhir_import_export.py [hunk 4]
```diff
@@ -103,115 +99,105 @@ def _can_create(output: dict,
...
-              project: str,
-              user: dict) -> bool:
-    """Check if user can read a project in the given program.
-
-    Args:
-        output: output dict the json that will be returned to the caller
-        program: program Gen3 program(-project)
-        project: project Gen3 (program-)project
-        user: user dict from arborist (aka profile)
+def _download_and_unzip(gh_username: str,
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: In the same way you have the [sower.DispatchArgs struct](https://github.com/calypr/forge/blob/fb409d2cf375ffb616bc986d2c4ee2537f64b094/client/sower/sower.go/#L38), I think you should define a similar class here to reference the association of the data passed via API

Comment (quinnwai) [reply etl-job/fhir_import_export.py]: In some ways this kinda starts at `_validate_and_extract_input` where you could do some arg level validation like you do before plopping it in a Python object and whisking it about

File: etl-job/fhir_import_export.py [hunk 5]
```diff
@@ -3,62 +3,58 @@
...
 
-def _input_data() -> dict:
-    """Get input data"""
-    assert 'INPUT_DATA' in os.environ, "INPUT_DATA not found in environment"
-    return json.loads(os.environ['INPUT_DATA'])
-
-
-def _get_program_project(input_data: dict) -> tuple:
-    """Get program and project from input_data"""
-    assert 'project_id' in input_data, "project_id not found in INPUT_DATA"
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py:63]: how do we plan on doing the cutover? Since these are breaking changes with no backward compatibility, we should have in writing our strategy for what to do if we ever need to self-service g3t as opposed to this newer ETL.

Comment (quinnwai) [reply etl-job/fhir_import_export.py:63]: Similar but separate, we get the following error

> [out] {"APIEndpoint":"https://caliper-training.ohsu.edu","bucketName":"cbds","file":null,"ghCommitHash":"6e05524027d0
22f8c3db9d887958950e0fb6188a","ghRepoUrl":"source.ohsu.edu/CBDS/forge-test.git","ghToken":"github_pat_11AAAADOY0EdulB
JBLDYRg_kZUomjtPau4wuAAAy1EL0ZT9e20Hg95IZWRuHbr3stdFNFB72OUJcSDsuek","ghUserName":"quinnwai","method":"put","profile"
:"local","projectId":"cbds-test_qww"}                                                                                
Traceback (most recent call last):                                                                                   
  File "/root/fhir_import_export.py", line 377, in <module>                                                          
    main()                                                                                                           
  File "/root/fhir_import_export.py", line 311, in main                                                              
    program, project = _get_program_project(input_data)                                                              
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                              
  File "/root/fhir_import_export.py", line 59, in _get_program_project                                               
    assert 'project_id' in input_data, "project_id not found in INPUT_DATA"                                          
           ^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                
AssertionError: project_id not found in INPUT_DATA

when pushing with an old image from forge. I imagine similar problems will happen the other way around (g3t push from forge). How do we ensure that for different versions of code, the users (and us) know that a change has occurred?

File: etl-job/fhir_import_export.py [hunk 6]
```diff
@@ -103,115 +99,105 @@ def _can_create(output: dict,
...
+            output=output,
+            access_token=_get_env_var('ACCESS_TOKEN')
+        )
+
+        for file in file_path.rglob('*'):
             if file.suffix in ['.ndjson', '.json']:
-                # output dictionary is capturing logs from this function
-                status = bulk_load_raw(hostname, _get_graphName(),
-                    f"{program}-{project}", str(file), output, _get_token())
+                status = bulk_load_raw(
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py:371]: running into a funky error:

did a forge publish (clean wd) on https://github.com/calypr/forge/issues/2

sower job
```
Created new ResearchStudy at /root/repo/test_qww/ResearchStudy.ndjson with ID 79e139e7-3526-4fe8-9123-9acf6c1dfe49   
Added new record: 7a844367-a92b-5892-900c-2731c6889a83                                                               
INFO:root:Added new record: 7a844367-a92b-5892-900c-2731c6889a83                                                     
Finished writing all records to files.                                                                               
INFO:root:Finished writing all records to files.                                                                     
Found files: ['/root/repo/test_qww/ResearchStudy.ndjson', '/root/repo/test_qww/DocumentReference.ndjson']            
INFO:root:Found files: ['/root/repo/test_qww/ResearchStudy.ndjson', '/root/repo/test_qww/DocumentReference.ndjson']  
bulk_load_raw return status {'error': 'service failure - try again later'}                                           
INFO:root:bulk_load_raw return status {'error': 'service failure - try again later'}    
```

grip deployment
```
$kc logs deployments/grip-deployment
{"Config":{"Server":{"HostName":"localhost","HTTPPort":"8201","RPCPort":"8202","WorkDir":"grip.work","ContentDir":"","ReadOnly":false,"EnablePlugins":false,"PluginDir":"","NoJobs":false,"Accounts":{"Auth":null,"Access":null},"DisableHTTPCache":true,"AutoBuildSchemas":false,"SchemaRefreshInterval":"24h0m0s","SchemaInspectN":500,"SchemaRandomSample":true,"RequestLogging":{"Enable":false,"HeaderWhitelist":["authorization","oauthemail","content-type","content-length","forwarded","x-forwarded-for","x-forwarded-host","user-agent"]}},"RPCClient":{"User":"","Password":"","ServerAddress":"localhost:8202","Timeout":"30s","MaxRetries":3},"Logger":{"Level":"info","Formatter":"text","OutputFile":"","JSONFormat":{"DisableTimestamp":false,"TimestampFormat":""},"TextFormat":{"ForceColors":false,"DisableColors":false,"DisableTimestamp":false,"TimestampFormat":"2006-01-02T15:04:05Z07:00","DisableSorting":false,"Indent":""}},"Default":"grids","Graphs":null,"Drivers":{"grids":{"Grids":"grip-grids.db","Badger":null,"Bolt":null,"Level":null,"Pebble":null,"MongoDB":null,"PSQL":null,"ExistingSQL":null,"Sqlite":null,"Gripper":null}},"Sources":{},"Kafka":{"Username":null,"Password":null,"Hostname":null,"Topic":null}},"level":"info","msg":"Starting Server","time":"2025-08-22T22:05:32Z"}
Default graph driver: grids
Method: (func(gripql.Client, map[string]string) (http.Handler, error))(0x4048784d80)
{"level":"info","msg":"Plugin gen3_writer.so loaded","time":"2025-08-22T22:05:32Z"}
{"level":"info","msg":"Plugin gql-gen.so loaded","time":"2025-08-22T22:05:32Z"}
Method: (func(gripql.Client, map[string]string) (http.Handler, error))(0x404cf12320)
{"level":"info","msg":"Plugin added to /writer/","time":"2025-08-22T22:05:32Z"}
{"level":"info","msg":"Plugin added to /graphql/","time":"2025-08-22T22:05:32Z"}
{"level":"info","msg":"TCP+RPC server listening on 8202","time":"2025-08-22T22:05:32Z"}
{"level":"info","msg":"HTTP proxy connecting to localhost:8201","time":"2025-08-22T22:05:32Z"}
fetching GRIDS graph CALYPR__schema__
2025/08/22 22:05:32 [JOB 1] WAL file grip-grids.db/CALYPR__schema__/000016.log with log number 000016 stopped reading at offset: 0; replayed 0 keys in 0 batches
{"data":200,"graph":"","level":"info","msg":"[200] healthy _status","status":200,"time":"2025-08-22T22:05:46Z"}
[GIN] 2025/08/22 - 22:05:46 | 200 |    3.812917ms |       10.42.0.1 | GET      "_status"
...
{"level":"info","msg":"Retrieved Cached token","time":"2025-08-22T22:06:05Z"}
{"level":"info","msg":"Resource List for method 'read': [/programs/HTAN_INT/projects/BForePC /programs/cbds/projects/045371b78e1849ac8723529c6610d537 ... /programs/cbds/projects/smmart_labkey_demo /programs/cbds/projects/test_qww ...]","time":"2025-08-22T22:06:05Z"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1688257]

goroutine 37630 [running]:
github.com/bmeg/grip/gdbi.(*DataElement).ToVertex(0x0)
	/go/pkg/mod/github.com/bmeg/grip@v0.0.0-20250714173925-c1892d4f63b1/gdbi/data_element.go:13 +0x17
github.com/bmeg/grip/engine/pipeline.Convert({0x32bcbb0, 0xc00175ced0}, 0xb0?, 0xc00446d6c0?, {0x32b5960?, 0xc009809bc0?})
	/go/pkg/mod/github.com/bmeg/grip@v0.0.0-20250714173925-c1892d4f63b1/engine/pipeline/pipes.go:163 +0x4b9
github.com/bmeg/grip/engine/pipeline.Run.func1()
	/go/pkg/mod/github.com/bmeg/grip@v0.0.0-20250714173925-c1892d4f63b1/engine/pipeline/pipes.go:112 +0x217
created by github.com/bmeg/grip/engine/pipeline.Run in goroutine 37629
	/go/pkg/mod/github.com/bmeg/grip@v0.0.0-20250714173925-c1892d4f63b1/engine/pipeline/pipes.go:103 +0xe5
```

**Review by quinnwai — CHANGES_REQUESTED** [review]
Thanks for these changes, this code looks very clean! Did an initial round of testing, added the comments for a couple bugs. Haven't looked too deep in the code except for debugging purposes so that I can still function as a fresh eyes  naive user.

For the next round of review, could you provide a testing checklist for the features and edge cases that I need to test? As well as the acceptance criteria for this PR? I'll be proactive during testing to find more edge cases but we should have consensus on some of it as we do testing.

Btw I've been testing on local so if I need anything specific add that to the testing notes. Rn I'm using
- sower: `fix_dataframer-new-ds`
- GRIP: `feature_indexing`
- and all the feature_rbac image tags already noted in [slack canvas](https://ohsucomputationalbio.slack.com/docs/T04MS9FKW/F0703QAPJK1)

File: etl-job/fhir_import_export.py (hunk 5 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py:63]: Cutover will happen when initializing a new project with the git-drs / forge / new etl pod stack. Currently we're still using g3t for publish, but once this gets tested a bit more then a cutover will have to be made

File: etl-job/fhir_import_export.py (hunk 6 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py:371]: I Think I have a fix for this. It will require you to delete your grip db and use this new grip image: 
```quay.io/ohsu-comp-bio/grip-caliper:feature_test-delete```
tested image is working on my machine, need to patch calypr development server next

File: etl-job/fhir_import_export.py (hunk 4 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: No sure what you're referring to here.

File: etl-job/fhir_import_export.py (hunk 3 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py:294]: what is the vibe ?

File: etl-job/fhir_import_export.py (hunk 5 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py:63]: We will have to manually migrate all projects to forge / git-drs. I suspect a few more git-drs PRs will be needed like to get git-drs error handling improved and  support for multiple calypr instances, before the migration will be made.

File: etl-job/fhir_import_export.py (hunk 4 above)
Comment (quinnwai) [reply etl-job/fhir_import_export.py]: oh just suggesting that you should pass in some sort of class instead of a lot of args. Thoughts?

Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: Yeah this could be refactored into class based approach like https://g.co/gemini/share/c7e6ac9b5e40. It's a little bit cleaner but still pretty messy. Parking lot it for the future perhaps.

File: etl-job/fhir_import_export.py (hunk 3 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: should be fixed now

File: etl-job/fhir_import_export.py [hunk 7]
```diff
@@ -140,246 +132,455 @@ def _can_read(output: dict,
...
+    logging.info("[out] retrieving user info...")
 
+    user = _get_user(auth)
     output = {'user': user['email'], 'files': [], 'logs': []}
-    # note, only the last output (a line in stdout with `[out]` prefix) is returned to the caller
-
-    # output['env'] = {k: v for k, v in os.environ.items()}
-
-    input_data = _input_data()
+    input_data = _parse_input_data()
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py:661]: I know we haven't historically done this, thoughts on loading input data into an object for reliability purposes?

is there any benefit?

File: etl-job/fhir_import_export.py [hunk 8]
```diff
@@ -140,246 +132,455 @@ def _can_read(output: dict,
...
+        if not _run_subprocess(["forge", "meta", validated_data['profile']], target_dir, output, f"ERROR RUNNING FORGE META INIT FOR PROJECT {program}-{project}"):
+            return False
+
+        for file in [f for f in os.listdir(os.path.join(target_dir, META_DIR)) if f.endswith(".ndjson")]:
+            meta_dir = os.path.join(META_DIR, file)
+            mv_cmd = ["mv", meta_dir, str(load_path)]
+            if not _run_subprocess(mv_cmd, target_dir, output, f"ERROR MOVING {file}"):
+                return False
+
+        # Nuke the whole ETL job if the config push doesn't work. -- controversial maybe not do this.
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py:613]: I haven't tested the config yet but I think a user probably wants a non-silent error if the config doesn't work. I'm with you that we should keep it for now

File: etl-job/fhir_import_export.py [hunk 9]
```diff
@@ -140,246 +132,455 @@ def _can_read(output: dict,
...
+    try:
+        grip_delete(
+            hostname,
+            graph_name=_get_env_var('GRIP_GRAPH_NAME'),
+            project_id=project_id,
+            output=output,
+            access_token=_get_env_var('ACCESS_TOKEN'),
+        )
+        output['logs'].append(f"EMPTIED graph for {project_id}")
+        for index in ["researchsubject", "specimen", "file"]:
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: any reason why we omit `medication_administration` and `group_member`? Also per guppy `values.yaml` should it be lower-cased?

```
guppy:
  image:
    repository: quay.io/ohsu-comp-bio/guppy-es7
    tag: "guppy-fix-refresh"
    pull_policy: Always
  enabled: true
  dbRestore: false
  indices:
  - index: specimen
    type: specimen
  - index: group_member
    type: group_member
  - index: research_subject
    type: research_subject
  - index: document_reference
    type: document_reference
  - index: medication_administration
    type: medication_administration
```

File: etl-job/fhir_import_export.py (hunk 7 above)
Comment (quinnwai) [reply etl-job/fhir_import_export.py:661]: PS: I did notice we do validation during the _put which is good

File: etl-job/fhir_import_export.py (hunk 4 above)
Comment (quinnwai) [reply etl-job/fhir_import_export.py]: yes makes sense

File: etl-job/fhir_import_export.py [hunk 10]
```diff
@@ -78,41 +71,40 @@ def _can_create(output: dict,
...
         if required_service not in user['authz']:
             output['logs'].append(f"{required_service} not found in user authz")
-            can_create = False
+            is_guppy_admin = False
         else:
-            if {'method': 'create', 'service': '*'} not in user['authz'][required_service]:
-                output['logs'].append(f"create not found in user authz for {required_service}")
-                can_create = False
+            if {'method': '*', 'service': "guppy"} not in user['authz'][required_service]:
+                output['logs'].append(f"'*' method not found in user authz for {required_service}")
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: how does a user get `/guppy_admin` permissions? I've run into errors during publish where it errors out but the job succeeds.

```
..."loading sqlite db...","Loaded cbds-git_drs_test","/guppy_admin not found 
in user resources","/guppy_admin not found in user authz","EMPTIED flat for cbds-git_drs_test"]} 
```

File: etl-job/fhir_import_export.py [hunk 11]
```diff
@@ -140,246 +132,455 @@ def _can_read(output: dict,
...
+        except IOError as err:
+            output["logs"].append(f"ERROR READING FILE {file}: {err}")
+            return False
+
+        headers = {
+            "Authorization": f"bearer {_get_env_var('ACCESS_TOKEN')}",
+            "Content-Type": "application/json"
+        }
+        try:
+            response = requests.put(f"{hostname}/ExplorerConfig/explorer/{base_name}", headers=headers, data=file_content)
```
Comment (quinnwai) [inline etl-job/fhir_import_export.py]: Are there authz checks in this call to determine whether someone has project access to that json? Do we need to do that here as well?

Ex: I know there's an `HTAN_INT-BForePC` project but no access to it in my `cbds-personal_proj` repo, so I manufacture + push an `HTAN_INT-BForePC.json` to change the config of a project I don't have access to

Comment (quinnwai) [reply etl-job/fhir_import_export.py]: Also I'm missing something, [gecko calls](https://github.com/ACED-IDP/gecko/pull/2/changes#diff-792360fa79f4d2b01996db5b1f2dbe613a7ac3d91b53caebad38d2de1675eac1L74) are a different endpoint than this, does this get routed through FEF?

File: etl-job/Dockerfile [hunk 12]
```diff
@@ -1,49 +1,57 @@
-# set base image (host OS)
+FROM golang:1.24.2-bookworm AS go-builder
+
+RUN go install github.com/calypr/forge@358d1fb6ec325f913bfd35a643885525fb01342f
+RUN go install github.com/calypr/git-drs@81a82438caa43f57a9c13158c4dded279698d2e8
```
Comment (quinnwai) [inline etl-job/Dockerfile]: shall we update this to [0dbe7e55d1f34435e25eadc319bf8bddf21b2083](https://github.com/calypr/git-drs/commit/0dbe7e55d1f34435e25eadc319bf8bddf21b2083) and then both of us retest?

File: etl-job/Dockerfile [hunk 13]
```diff
@@ -1,49 +1,57 @@
-# set base image (host OS)
+FROM golang:1.24.2-bookworm AS go-builder
+
+RUN go install github.com/calypr/forge@358d1fb6ec325f913bfd35a643885525fb01342f
```
Comment (quinnwai) [inline etl-job/Dockerfile]: update to [84cc9ed6f56372d3e9eb6c729be3f1da15c0ed1d](https://github.com/calypr/forge/pull/5/changes/84cc9ed6f56372d3e9eb6c729be3f1da15c0ed1d)?

**Review by quinnwai — DISMISSED** [review]
Nicely organized! Some minor comments and a request to update + retest.

My biggest comment overall tho is that it will be a headache if BForePC updates can't be done using the old tag, I get that this is a hard cutover but to cover our grounds in case HTAN users need a change to happen rapidly and we can't fix it in time.

File: etl-job/fhir_import_export.py (hunk 11 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: yes it's handled in nginx and the frontend. Permissions checking is done by checking create permissions for hte resource path that you are asking for, similar to how write checks are done in the etl pod, except it is done at the Server level in gecko.

File: etl-job/Dockerfile (hunk 12 above)
Comment (matthewpeterkort) [reply etl-job/Dockerfile]: probably a good idea

File: etl-job/fhir_import_export.py (hunk 10 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: Yeah that is a special permission that is added. the user.yaml probably needs to get updated, but calypr_admin util should already be setup for it https://github.com/calypr/admin/blob/development/calypr_admin/admin/collaborator/access/policies/add-user-guppy-admin.yaml

File: etl-job/fhir_import_export.py (hunk 9 above)
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py]: yes to both that's just old code range over INDEX_NAMES instead

**Comment by matthewpeterkort** [comment]
@codex review

File: etl-job/fhir_import_export.py [hunk 14]
```diff
@@ -6,70 +6,71 @@
...
 
-def _get_program_project(input_data: dict) -> tuple:
-    """Get program and project from input_data"""
-    assert 'project_id' in input_data, "project_id not found in INPUT_DATA"
-    assert '-' in input_data['project_id'], 'project_id must be in the format <program>-<project>'
-    return input_data['project_id'].split('-')
+def _get_program_project(input_data: Dict[str, Any]) -> Tuple[str, str]:
+    """Extract program and project from input_data."""
+    project_id = input_data.get("projectId")
+    if not project_id or "-" not in project_id:
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply etl-job/fhir_import_export.py:70]: no this is breaking change and is reflected in the client

**Review by bwalsh — APPROVED** [review]
